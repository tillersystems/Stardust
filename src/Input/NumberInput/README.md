# NumberInput

### Usage

```jsx
import { NumberInput } from '@tillersystems/stardust';
```

<!-- STORY -->

<!-- PROPS -->

### Example

#### Standard use case

The following snippet shows how to create a controlled input.

```jsx
import { NumberInput } from '@tillersystems/stardust';

render() {
  return (
    <NumberInput
      value={myValue}
      onChange={(newValue) => handleChange(newValue)}
    />
  );
}
```

#### Disabled input

The following snippet shows how to create a disabled input. _Note that even if the user cannot
enter values by focusing on the input, using a controlled input would allow use to change its
value_.

```jsx
import { NumberInput } from '@tillersystems/stardust';

render() {
  return (
    <NumberInput disabled />
  );
}
```

#### NumberInput with label

The following snippets show how to create an input with a label.

```jsx
import { NumberInput } from '@tillersystems/stardust';

render() {
  return (
    <NumberInput label={{ icon: "euro" }} labelPosition="right" />
  );
}
```

```jsx
import { NumberInput } from '@tillersystems/stardust';

render() {
  return (
    <NumberInput label={{ text: "http://" }} labelPosition="right" />
  );
}
```
