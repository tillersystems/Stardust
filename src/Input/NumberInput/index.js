import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import TextInput from '../TextInput';

/**
 * Gets the locale decimal separator character from the browser.
 *
 * @returns {string} The local decimal separator.
 */
function getLocaleSeparator() {
  return (1.1).toLocaleString().substring(1, 2);
}

function escapeSeparator(separator) {
  return separator === '.' ? '\\.' : separator;
}

class NumberInput extends PureComponent {
  /** Display name. */
  static displayName = 'NumberInput';

  /** Prop types validation. */
  static propTypes = {
    // - - - Input related props - - -
    value: PropTypes.number,
    onChange: PropTypes.func,

    // - - - Appearance related props - - -
    validate: PropTypes.bool,

    // - - - Number input related props - - -
    min: PropTypes.number,
    max: PropTypes.number,
    step: PropTypes.number,

    // - - - Number format related props - - -
    decimals: PropTypes.number,
    separator: PropTypes.string,
  };

  /** Default props. */
  static defaultProps = {
    value: null,
    onChange: () => {},

    validate: false,

    min: null,
    max: null,
    step: 1,
    decimals: 0,
    separator: getLocaleSeparator(),
  };

  /** Internal state. */
  state = {
    // Raw value: This is the string used to control the inner input.
    rawValue: null,

    // Parsed value: This is the parsed number value from the raw value.
    parsedValue: null,

    // Whether the component has focus. This value is used to allow keyboard increase and decrease
    // of the value.
    hasFocus: false,
  };

  /**
   * Handles mounting step in component's lifecycle.
   */
  componentDidMount() {
    document.addEventListener('keydown', this.handleKeyDown);
  }

  /**
   * Handles unmounting step in component's lifecycle.
   */
  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyDown);
  }

  /**
   * Formats the given value into a displayable string in the inner TextInput.
   *
   * @returns {string}
   */
  formatValue = () => {
    const { rawValue, parsedValue } = this.state;
    const { value, decimals, separator } = this.props;

    const displayedValue = value || parsedValue;

    return (displayedValue
      ? displayedValue.toFixed(decimals)
      : rawValue
        ? rawValue
        : (0).toFixed(decimals)
    ).replace('.', separator);
  };

  /**
   * Checks whether the given value is a string of a number or not.
   *
   * @param {string} rawValue - The string to check.
   *
   * @return {boolean} true if the given string is a number, false otherwise.
   */
  isNumber = rawValue => {
    const { separator } = this.props;
    return new RegExp(`^\\d+(${escapeSeparator(separator)}\\d+)?$`).test(rawValue);
  };

  /**
   * Updates the value of the component.
   * This method updates the parsed value in the component's internal state. In addition it checks
   * for validity of the new value; in case it is out of the configured bounds, it modifies it.
   *
   * @param {number} newParsedValue - The new value.
   */
  updateValue = newParsedValue => {
    const { onChange, min, max } = this.props;

    // Checks bounds.
    // Have to use value != null since 0 is equivalent to null boolean wise.
    let parsedValue = newParsedValue;
    if (min != null && newParsedValue < min) {
      parsedValue = min;
    } else if (max != null && newParsedValue > max) {
      parsedValue = max;
    }

    this.setState({ ...this.state, parsedValue }, () => {
      onChange(parsedValue);
    });
  };

  /**
   * Handles change in the inner TextInput.
   *
   * @param {string} rawValue - The new value of the inner TextInput.
   */
  handleChange = rawValue => {
    const { separator } = this.props;

    const parsedValue = this.isNumber(rawValue)
      ? parseFloat(rawValue.replace(separator, '.'))
      : null;

    this.setState({ ...this.state, rawValue }, () => this.updateValue(parsedValue));
  };

  /**
   * Handles focus on the inner TextInput.
   */
  handleFocus = () => {
    this.setState({ ...this.state, hasFocus: true });
  };

  /**
   * Handles blur on the inner TextInput.
   */
  handleBlur = () => {
    this.setState({ ...this.state, hasFocus: false });
  };

  /**
   * Handles keyboard's key down event.
   *
   * @param {KeyEvent} event - The keyboard event.
   */
  handleKeyDown = event => {
    const { hasFocus, parsedValue } = this.state;
    const { step } = this.props;

    // Checks if the input has focus since we connect the event listener to the window.
    if (hasFocus) {
      if (event.code === 'ArrowUp') {
        this.setState({ ...this.state, rawValue: this.formatValue() }, () => {
          this.updateValue(parsedValue + step);
        });
      }

      if (event.code === 'ArrowDown') {
        this.setState({ ...this.state, rawValue: this.formatValue() }, () => {
          this.updateValue(parsedValue - step);
        });
      }
    }
  };

  /**
   * Renders the component.
   *
   * @returns {jsx}
   */
  render() {
    const { rawValue } = this.state;
    const { validate } = this.props;

    const hasValidValue = this.isNumber(rawValue);

    return (
      <TextInput
        value={this.formatValue()}
        success={validate && hasValidValue}
        error={validate && !hasValidValue}
        onChange={this.handleChange}
        onFocus={this.handleFocus}
        onBlur={this.handleBlur}
      />
    );
  }
}

export default NumberInput;
