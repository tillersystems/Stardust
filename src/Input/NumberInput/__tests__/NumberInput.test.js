import React from 'react';
import 'jest-styled-components';

import NumberInput from '..';

describe('<NumberInput />', () => {
  it('should render withouth a problem', () => {
    const render = mountWithTheme(<NumberInput />);
    expect(render).toMatchSnapshot();
  });

  it('should render withouth a problem when focused and unfocused', () => {
    const render = mountWithTheme(<NumberInput />);

    render.find('input').simulate('focus');
    expect(render).toMatchSnapshot();

    render.find('input').simulate('blur');
    expect(render).toMatchSnapshot();
  });
});
