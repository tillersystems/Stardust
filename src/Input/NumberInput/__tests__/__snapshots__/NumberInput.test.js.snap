// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NumberInput /> should render withouth a problem 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 25rem;
  height: 4rem;
  overflow: hidden;
  border-radius: 0.4rem;
  border: 1px solid hsl(218,18%,91%);
  background-color: hsl(0,100%,100%);
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  background-color: hsl(0,100%,100%);
  margin: 0 0.7rem;
  border-radius: 0.4rem;
  font-size: 1.2rem;
  border: none;
  padding: 0;
  text-align: left;
}

.c1:focus,
.c1:active {
  outline: none;
}

<NumberInput
  decimals={0}
  max={null}
  min={null}
  onChange={[Function]}
  separator="\\\\."
  step={1}
  validate={false}
>
  <TextInput
    _onBlur={[Function]}
    _onFocus={[Function]}
    disabled={false}
    error={false}
    fluid={false}
    id=""
    info={false}
    label={null}
    labelPosition="left"
    loading={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    password={false}
    placeHolder=""
    search={false}
    success={false}
    tabIndex="0"
    value="0"
    warning={false}
    width="25rem"
  >
    <styled.div
      disabled={false}
      error={false}
      hasFocus={false}
      info={false}
      success={false}
      warning={false}
      width="25rem"
    >
      <div
        className="c0"
        disabled={false}
        width="25rem"
      >
        <styled.input
          disabled={false}
          id=""
          innerRef={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          tabIndex="0"
          type="text"
          value="0"
        >
          <input
            className="c1"
            disabled={false}
            id=""
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder=""
            tabIndex="0"
            type="text"
            value="0"
          />
        </styled.input>
      </div>
    </styled.div>
  </TextInput>
</NumberInput>
`;
